gem/optimise.py

```diff
@@ -413,3 +475,555 @@ def aggressive_unroll(expression):
+    """
+    arg_ind_flat = tuple([i for indices in arg_ind for i in indices])
+    multiindex = ()
+    if isinstance(node, IndexSum):
+        multiindex = node.multiindex
+        node = node.children[0]
+    node = expand_products(node, arg_ind_flat)
+    summands = collect_terms(node, Sum)
+    rep = []
+    for summand in summands:
+        d = Summand(arg_ind_flat=arg_ind_flat)
+        for factor in collect_terms(summand, Product):
+            key = decide_key(factor, arg_ind_flat, set(arg_ind_flat))
+            d.setdefault(key, []).append(factor)
+        rep.append(d)
+    return (rep, multiindex)
+
+
+def optimise(node, quad_ind, arg_ind):
+    if isinstance(node, Constant):
+        return node
+    rep, multiindex = build_repr(node, arg_ind)
+    lo = LoopOptimiser(rep=rep, multiindex=multiindex, arg_ind=arg_ind)
+    optimal_arg = lo.find_optimal_arg_factors()
+    lo.factorise_arg(optimal_arg[0] + optimal_arg[1])
+
+    print('Node1 = {0}'.format(count_flop(lo.node1)))
+    print('Node2 = {0}'.format(count_flop(lo.node2)))
+    if count_flop(lo.node1) < count_flop(lo.node2):
+        return lo.node1
+    else:
+        return lo.node2
+
+
+def optimise_expressions(expressions, quadrature_indices, argument_indices):
+    if propagate_failure(expressions):
+        return expressions
+    return [optimise(node, quadrature_indices, argument_indices) for node in expressions]
+
+
+def propagate_failure(expressions):
+    """
+    Check if any gem nodes is Failure. In that case there is no need for subsequent optimisation.
+    """
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return True
+    return False
+
+
+def _list_2_node(children, self):
+    if not self.balanced:
+        return reduce(self.func, children, self.base)
+    if len(children) < 3:
+        return reduce(self.func, children, self.base)
+    else:
+        mid = len(children) // 2
+        return self((self(children[:mid]), self(children[mid:])))
+
+
+def list_2_node(function, children, balanced=True, sort=False):
+    """
+    generate gem node from list of children
+    use recursion so that each term is not too long
+    i.e. (a+b) + (c+d) instead of (((a+b)+c)+d
+    :param children: list of children nodes
+    :return: gem node
+    """
+    # TODO: DAG awareness. Hashing with tuple is probably slow here.
+    if sort:
+        children = sorted(children)
+    if function == Sum:
+        base = Zero()
+    elif function == Product:
+        base = one
+    else:
+        raise AssertionError('Cannot combine unless Sum or Product')
+    mapper = Memoizer(_list_2_node)
+    mapper.func = function
+    mapper.base = base
+    mapper.balanced = balanced
+    return mapper(tuple(children))
+
+
+def sort_keys(keys, arg_ind_flat):
+    """
+    sort keys of :class: `Summand` by lexicographical order, according to the order they
+    appear in :param arg_ind_flat.
+    e.g. [j, k, (j,k)] => [j, (j,k), k]
```

Comment:
This is ugly... if you don't have backwards compatibility constraints, why would you torture yourself with inconsistent API design? If entry isn't always a single index, just make it always a tuple. You will spare yourself some (potentially many) stupid branching. (Is this an `Index`? Is this a `tuple`?)